cpp/src/arrow/compute/kernels/scalar_validity.cc

```diff
@@ -189,6 +191,72 @@ Status ConstBoolExec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
   return Status::OK();
 }
 
+struct NonZeroVisitor {
+   UInt32Builder *builder;
+   const ArrayData& array;
+
+   NonZeroVisitor(UInt32Builder *builder, const ArrayData& array)
+   : builder(builder), array(array) {}
+
+  Status Visit(const DataType& type) {
+    return Status::TypeError("Unsupported type for nonzero: ", type.ToString());
```

Comment:
[Use `Status::NotImplemented` as in the default TypeVisitors.](https://github.com/edponce/arrow/blob/master/cpp/src/arrow/visitor.cc#L81)

Reply:
:+1:

Reply:
:+1:, used `enable_if_t<is_primitive_ctype<Type>::value, Status>` so that if in the future support for additional types has to be added they can easily be `||` with the current condition.